Reusable blocks: Use view context, always include title.raw and content.raw

noisysocks opened the pull request:
Supersedes https://github.com/WordPress/gutenberg/pull/11891. Fixes #11343.

Existing reusable blocks on a post would not render if the user was an author or a contributor. This happened because requests to fetch a single block or post are blocked when `?context=edit` is passed and the current user is not an editor.

The fix is to stop passing `?context=edit` to block requests, but this causes another problem which is that Gutenberg needs access to both `title.raw` and `content.raw` in the API response.

We therefore both stop passing `?context=edit` to block requests *and* modify the blocks API so that `title.raw` and `content.raw` are always provided.

#### How to test

1. Create two users, `user1` (any role) and `user2` (author role)
1. Log in as `user1` and create a reusable block
1. Log in as `user2` and create a post using `user1`'s block
1. The block should render correctly.
1. Save the post and refresh the edit screen.
1. The block should again render correctly.

noisysocks committed:
Reusable blocks: Use view context

Existing reusable blocks on a post would not render if the user was an
author or a contributor. This happened because requests to fetch a
single block or post are blocked when ?context=edit is passed and the
current user is not an editor.

The fix is to stop passing `?context=edit` to block requests, but this
causes another problem which is that Gutenberg needs access to both
`title.raw` and `content.raw` in the API response.

We therefore both stop passing `?context=edit` to block requests *and*
modify the blocks API so that `title.raw` and `content.raw` are always
provided.

Lastly, we modify the blocks API to always omit `content.rendered`, as
it doesn't make sense for a reusable block to have rendered content on
its own, since rendering a block requires it to be inside a post or a
page.

noisysocks commented:
Wasn't able to re-open #11891 so this is a new PR to replace it.

I updated the inline comment to explain how `unset()` achieves what we want, and have tweaked the Reusable Block's description in the sidebar with an aim to better explain that blocks can be seen and used by non-editors:

<img width="294" alt="screen shot 2018-11-20 at 11 45 22" src="https://user-images.githubusercontent.com/612155/48744046-c8fe1600-ecb9-11e8-9d2a-e50519bf902f.png">

cc. @mtias @pento @peterwilsoncc

kadamwhite commented:
The restrictions around .raw were implemented for security purposes; it is possible to have content within the raw versions of title and (more commonly) content that would leak potentially sensitive data using this approach. This PR therefore changes the WordPress permissions model.  I understand the constraints of shared blocks which suggest an approach like this but I cannot endorse sidestepping the mindfully-designed permissions checks that restrict these fields.

pento commented on lib/class-wp-rest-blocks-controller.php:
```diff
@@ -33,4 +33,28 @@ public function check_read_permission( $post ) {
+	public function get_item_schema() {
+		$schema = parent::get_item_schema();
+
+		// To use a reusable block in the editor, one needs access to its raw markup.
```
Multiline comments should be in long comment form:

```php
/*
 * foo
 * bar
 */
```

noisysocks replied:
👍 done.

pento commented on phpunit/class-rest-blocks-controller-test.php:
```diff
@@ -176,4 +176,35 @@ public function test_capabilities( $action, $role, $expected_status ) {
+	 * low-privileged role when no `context` param is provided.
+	 */
+	public function test_includes_raw_title_and_content() {
+		$user_id = $this->factory->user->create( array( 'role' => 'author' ) );
```
User creation (and deletion at the end of this function) should be in the `::wpSetUpBeforeClass()` and `::wpTearDownAfterClass()` methods.

If the `::assertEquals()` call fails it will throw an exception, causing the test method to exit immediately, so the user wouldn't be properly deleted.

noisysocks replied:
I've moved user deletion to `::wpTearDownAfterClass`. The user we require is different for every test case so I've kept user creation as is.

noisysocks replied:
Oops, the fake user being different for every test case means we should be deleting it in `::tearDown()`, not `::wpTearDownAfterClass()`. Fixed in f8ebe5f72.

pento replied:
It's fine to create a bunch of users in the setup and teardown methods, even when they're only used in one test. Kind of like what we do in [`WP_REST_Themes_Controller_Test`](https://github.com/WordPress/gutenberg/blob/master/phpunit/class-wp-rest-themes-controller-test.php).

It's easier to read if those methods are balanced, too: teardown just undoes whatever setup did.

noisysocks replied:
👍 changed in 4394eeb99.

pento reviewed (commented):
Given that we should really only ever need the `raw` post data, and not the `rendered` post data, I'm inclined to think that we should remove `rendered` from the returned data. This will help to avoid the confusion from an API perspective.

noisysocks commented on lib/class-wp-rest-blocks-controller.php:
```diff
@@ -33,4 +33,28 @@ public function check_read_permission( $post ) {
+		// because its schema specifies `'context' => array( 'edit' )`. Here, we use
+		// `unset()` to modify the schema so that the properties do not have this
+		// `context` restriction.
+		unset( $schema['properties']['title']['properties']['raw']['context'] );
```
👍 makes it a lot easier to understand, too.

noisysocks commented:
> Given that we should really only ever need the raw post data, and not the rendered post data, I'm inclined to think that we should remove rendered from the returned data. This will help to avoid the confusion from an API perspective.

I've made this change and added some comments explaining the rationale.

noisysocks committed:
Delete fake user after each test case, not after all test cases

pento commented on lib/class-wp-rest-blocks-controller.php:
```diff
@@ -33,4 +33,55 @@ public function check_read_permission( $post ) {
+	public function filter_response_by_context( $data, $context ) {
+		$data = parent::filter_response_by_context( $data, $context );
+
+		/**
```
```suggestion
		/*
```

🙂

pento committed:
Typo: /** → /*

Co-Authored-By: noisysocks <robert@noisysocks.com>

pento commented on lib/class-wp-rest-blocks-controller.php:
```diff
@@ -33,4 +33,55 @@ public function check_read_permission( $post ) {
+		 * reusable block to have rendered content on its own, since rendering a
+		 * block requires it to be inside a post or a page.
+		 */
+		unset( $schema['properties']['content']['properties']['rendered'] );
```
Do we need to keep the rendered title?

noisysocks committed:
Create all fake data before any test runs

noisysocks committed:
Omit `title.rendered` from the blocks REST API

The rendered title isn't used, so we may as well remove it so that
`content` and `title` are aligned.

noisysocks commented:
I removed `title.rendered` in 55c7800ed. It's not very relevant in the case of blocks since we don't ever render a block by itself on the front-end.

gziolo commented:
What about `contributor` role? I did the following:

1. Create a post as a contributor.
2. Added a paragraph.
3. Saved as a draft.
4. Switched to the admin role.
5. Turned the existing paragraph into a reusable block.
6. Saved updates.
7. Logged into the contributor account.
8. I still can edit my post, but reusable blocks don't load at all.

![screen shot 2018-11-20 at 08 03 16](https://user-images.githubusercontent.com/699132/48756880-2e8eda00-ec9b-11e8-95a8-2e993493108c.png)

gziolo commented:
I can confirm it works perfectly fine for the `editor` role.

gziolo commented:
```<!-- wp:block {"ref":47} /-->

<!-- wp:block /-->```

I turned paragraph block into a reusable one when using admin role. It looks like you can't load existing reusable blocks as a contributor.

gziolo committed:
Fix fetching when using reusable blocks as a contributor

gziolo commented:
I added a fix for the contributor role in b0333f3.

I have no idea whether it is valid or not. @youknowriad I would appreciate your feedback.

gziolo reviewed (approved):
I added the last commit to fix the issue for the contributor role. I would appreciate a sanity check from someone else and we can 🚢 

Let's work on UI improvements in the follow-up PRs.

noisysocks committed:
Fix reusable block effect tests

noisysocks committed:
Fix reusable block E2E tests
